Add 356 release notes (#7757)
trinodb/trino: docs/src/main/sphinx/release/release-356.md

> # Release 356 (XX Apr 2021)

Reviewer: Don't forget to fill in the date.

* Add dedicated access control for creating and dropping materialized views. ({issue}`7645`)
* Add dedicated access control for refreshing materialized views. Insert privilege on storage table is no
  longer required. ({issue}`7707`)
* Fix authentication failure when providing multiple scope values for `http-server.authentication.oauth2.scopes`. ({issue}`7706`)

## JDBC driver

* Add support for caching OAuth2 credentials in memory to avoid unnecessary authentication flows. ({issue}`7309`)

## BigQuery connector

* Add support for `CREATE SCHEMA` and `DROP SCHEMA` statements. ({issue}`7543`)
* Improve table listing performance when case insensitive matching is enabled. ({issue}`7628`)

## Cassandra connector

* Fix `NullPointerException` when reading an empty timestamp value. ({issue}`7433`)

## Hive connector

* Improve performance when reading dictionary-encoded Parquet files. ({issue}`7754`)
* Fix incorrect results when referencing nested fields with non-lowercase names from ORC files. ({issue}`7350`)
* Allow reading from ORC ACID transactional tables when `_orc_acid_version` metadata files are missing. ({issue}`7579`)

> * Add support for connection proxying for Azure ADL endpoints. ({issue}`7509`)

Reviewer 2: ADL -> ADLS

## MemSQL connector

* Improve metadata caching hit rate. ({issue}`7039`)

## MongoDB connector

* Fix handling of non-lowercase MongoDB views. ({issue}`7491`)

## MySQL connector

* Improve metadata caching hit rate. ({issue}`7039`)
* Exclude an internal `sys` schema from schema listings. ({issue}`6337`)

## Oracle connector

* Improve metadata caching hit rate. ({issue}`7039`)

## Phoenix connector

* Improve metadata caching hit rate. ({issue}`7039`)

## PostgreSQL connector

* Improve metadata caching hit rate. ({issue}`7039`)
* Cancel query on PostgreSQL when the Trino query is cancelled. ({issue}`7306`)
* Discontinue support for PostgreSQL 9.5, which has reached end of life. ({issue}`7676`)

## Redshift connector

* Improve metadata caching hit rate. ({issue}`7039`)

## SQL Server connector

* Improve metadata caching hit rate. ({issue}`7039`)
* Fix query failure when snapshot isolation is disabled in target SQL Server database,
  but `READ_COMMITTED_SNAPSHOT` is still enabled. ({issue}`7548`)
* Fix reading `date` values before 1583-10-14. ({issue}`7634`)

## SPI

* Require that `ConnectorMaterializedViewDefinition` provides a view owner. ({issue}`7489`)

> * Add `Connector#getMaterializedViewPropertyManager` for specifying materialized view properties. ({issue}`7615`)

Reviewer 2: We normally write these as `Connector.getMaterializedViewPropertyManager()`

> # Release 356 (XX Apr 2021)

Reviewer: nit: some of the items have `.` at the end and some do not.

## General

* Add support for {doc}`/sql/match-recognize`. ({issue}`6111`)
* Add {func}`soundex` function. ({issue}`4022`)  
* Introduce `system.metadata.materialized_view_properties` table for listing available materialized view properties. ({issue}`7615`)
* Introduce planning time limit which can be specified using `query.max-planning-time` config property ({issue}`7213`)
* Improve cancellation of queries during planning phase ({issue}`7213`)
* Improve performance of `ORDER BY ... LIMIT` queries over a `LEFT JOIN`. ({issue}`7028`)
* Improve performance of queries with predicates on boolean columns ({issue}`7263`)
* Improve planning time for queries with large `IN` predicates. ({issue}`7556`)
* Improve performance of queries that contain join on `VARCHAR` keys of different length. ({issue}`7644`)
* Reduce coordinator network overhead when scheduling queries. ({issue}`7351`)

> * Fix join spill deadlock. ({issue}`7455`)

Reviewer: nit: Maybe `Fix possible deadlock for JOIN queries when spilling is enabled`? `Fix possible deadlock for JOIN queries which spill to disk`?

> * Fix invalid query result for queries with full join which has scalar (left or right) input. ({issue}`7629`)

Reviewer: nit: `s/invalid/incorrect/`

* Add dedicated access control for creating and dropping materialized views. ({issue}`7645`)

## JDBC driver

* Add `externalAuthenticationTokenCache=MEMORY` property so Oauth2 credentials are cached in memory. Thanks to this user can avoid unnecessary authentication flow. ({issue}`7309`)

## BigQuery connector

* Add support for `CREATE SCHEMA` and `DROP SCHEMA` statements. ({issue}`7543`)
* Improve performance for listing tables when case insensitive matching (`bigquery.case-insensitive-name-matching`) is enabled. ({issue}`7628`)

## Cassandra connector

> * Fix `NullPointerException` when reading an empty timestamp value. ({issue}`7433`)

Reviewer: nit: `s/empty/NULL/`?

Reviewer 3: I think no. If user stored `NULL` in Cassandra, this issue doesn't happen.

## Hive connector

> * Fix incorrect results when non-lowercase field dereference is pushed into the ORC reader,
>   if ORC file schema has nested fields with a different case. ({issue}`7350`)

Reviewer: ```suggestion
* Fix incorrect results when non-lowercase field dereference is pushed into the ORC reader,
  if ORC file schema has nested fields with non-lowercase names. ({issue}`7350`)
```

IMO `different case` is not straightforward - different from what? The table definition (which is always lowercase IIRC)?

* Allow reading from ORC ACID transactional tables when `_orc_acid_version` metadata file is missing. ({issue}`7579`)
* Allow specifying Azure ADL connection proxy via `hive.azure.adl-proxy-host` config property. ({issue}`7509`)

## Iceberg connector

* Show Iceberg tables created by other engines in ``SHOW TABLES`` output. ({issue}`1592`)
* Fix querying materialized views that were created using session catalog and schema. ({issue}`7711`)

## Kafka connector

> * Add support for SSL security protocol. ({issue}`6929`

Reviewer: ```suggestion
* Add support for SSL security protocol. ({issue}`6929`)
```

## MongoDB connector

* Fix handling of non-lowercase MongoDB views. ({issue}`7491`)

## PostgreSQL connector

> * Cancel PostgreSQL remote query on when Trino query is cancelled. ({issue}`7306`)

Reviewer: ```suggestion
* Cancel query on PostgreSQL when the Trino query is cancelled. ({issue}`7306`)
```

# Release 356 (XX Apr 2021)

## General

* Add support for {doc}`/sql/match-recognize`. ({issue}`6111`)
* Add {func}`soundex` function. ({issue}`4022`)  
* Introduce `system.metadata.materialized_view_properties` table for listing available materialized view properties. ({issue}`7615`)

> * Introduce planning time limit which can be specified using `query.max-planning-time` config property. ({issue}`7213`)

Author: How about:

> Add support for limiting the maximum planning via the `query.max-planning-time` configuration property.

* Add configuration for the query info sent to clients.  This allows clients to be directed to a different location for 
  information about their query.  This is set using the `query.info-url-template` property.  ({issue}`7678`)
* Improve cancellation of queries during planning phase. ({issue}`7213`)

> * Improve performance of `ORDER BY ... LIMIT` queries over a `LEFT JOIN`. ({issue}`7028`)

Author: > but we chose Top N pushdown as name for such clauses

No, use use `ORDER BY ... LIMIT` for that. Top N doesn't mean anything in SQL -- it's just an implementation detail of the engine and internal APIs. Referring to it as `ORDER BY ... LIMIT` is the right thing to put it in context of the SQL someone might write.

> * Improve performance of `ORDER BY ... LIMIT` queries over a `LEFT JOIN`. ({issue}`7028`)

Reviewer 2: We call it Top-N pushdown in https://github.com/trinodb/trino/blob/master/docs/src/main/sphinx/optimizer/pushdown.rst#top-n-pushdown to have a convenient way to refer to it, since trying to use SQL syntax in all the links doesn't really work. That documentation explains why it is called that and gives both syntax forms.

However, for this release note, it's actually not about pushdown, so using the syntax makes sense.

* Improve performance of queries with predicates on boolean columns. ({issue}`7263`)
* Improve planning time for queries with large `IN` predicates. ({issue}`7556`)
* Improve performance of queries that contain join on `VARCHAR` keys of different length. ({issue}`7644`)
* Improve performance of queries with `late_materialization` enabled. ({issue}`7695`)
* Reduce coordinator network overhead when scheduling queries. ({issue}`7351`)
* Fix possible deadlock for `JOIN` queries when spilling is enabled. ({issue}`7455`)
* Fix incorrect results for queries with full join which has scalar (left or right) input. ({issue}`7629`)
* Add dedicated access control for creating and dropping materialized views. ({issue}`7645`)

## JDBC driver

> * Add `externalAuthenticationTokenCache=MEMORY` property so Oauth2 credentials are cached in memory. Thanks to this user can avoid unnecessary authentication flow. ({issue}`7309`)

Author: How about:

> Add support for caching OAuth2 credentials in memory to avoid unnecessary authentication flows.

Generally, we want to explain what's the impact of the change.

> * Add `externalAuthenticationTokenCache=MEMORY` property so Oauth2 credentials are cached in memory. Thanks to this user can avoid unnecessary authentication flow. ({issue}`7309`)

Reviewer: People generally want to know how to enable the change. So if we drop the config property we should still link to the config property somehow (unless it's enabled by default).

## BigQuery connector

* Add support for `CREATE SCHEMA` and `DROP SCHEMA` statements. ({issue}`7543`)
* Improve performance for listing tables when case insensitive matching (`bigquery.case-insensitive-name-matching`) is enabled. ({issue}`7628`)

## Cassandra connector

* Fix `NullPointerException` when reading a null timestamp value. ({issue}`7433`)

## Hive connector

* Improve performance when reading dictionary-encoded Parquet files. ({issue}`7754`)
* Fix incorrect results when non-lowercase field dereference is pushed into the ORC reader,

>   if ORC file schema has nested fields with non-lowercase names. ({issue}`7350`)

Reviewer: I think it's important - removing it makes it sound like the ORC reader was always incorrect with nested fields while that's not the case.

* Allow reading from ORC ACID transactional tables when `_orc_acid_version` metadata file is missing. ({issue}`7579`)
* Allow specifying Azure ADL connection proxy via `hive.azure.adl-proxy-host` config property. ({issue}`7509`)

## Iceberg connector

* Show Iceberg tables created by other engines in `SHOW TABLES` output. ({issue}`1592`)
* Improve performance when reading dictionary-encoded Parquet files. ({issue}`7754`)
* Fix querying materialized views that were created using session catalog and schema. ({issue}`7711`)

## Kafka connector

* Add support for SSL security protocol. ({issue}`6929`)

## MongoDB connector

* Fix handling of non-lowercase MongoDB views. ({issue}`7491`)

## PostgreSQL connector

* Cancel query on PostgreSQL when the Trino query is cancelled. ({issue}`7306`)

> * Discontinue support for PostgreSQL 9.5.

Reviewer 3: How about mentioning issue number? 
```suggestion
* Discontinue support for PostgreSQL 9.5 because of EOL. ({issue}`7676`)
```

## SQL Server connector

* Fix query failure when snapshot isolation (`ALLOW_SNAPSHOT_ISOLATION`) is disabled in target SQL Server database, but

>   but `READ_COMMITTED_SNAPSHOT` is still enabled. ({issue}`7548`)

Reviewer 3: nit: Fix repeated `but but`

* Fix reading `date` values before 1583-10-14. ({issue}`7634`)

## Other connectors

> This applies to the MemSQL, MySQL, Oracle, Phoenix, PosgreSQL, Redshift, and SQL Server connectors:

Author: We've done this many times, depending on whether it affects multiple connectors and there are no other items to document

* Improve metadata caching hit rate by sharing cache entries between different session users. ({issue}`7039`)

## SPI

> * Require that ``ConnectorMaterializedViewDefinition`` provides a view owner. ({issue}`7489`)

Reviewer 3: ```suggestion
* Require that `ConnectorMaterializedViewDefinition` provides a view owner. ({issue}`7489`)
```

* Add support for {doc}`/sql/match-recognize`. ({issue}`6111`)
* Add {func}`soundex` function. ({issue}`4022`)  
* Introduce `system.metadata.materialized_view_properties` table for listing available materialized view properties. ({issue}`7615`)

> * Add support for limiting the maximum planning via the `query.max-planning-time` configuration property. ({issue}`7213`)

Reviewer: ```suggestion
* Add support for limiting the maximum planning time via the `query.max-planning-time` configuration property. ({issue}`7213`)
 ```
